pkg/cmd/experimental/diagnostics/cluster.go

```diff
@@ -0,0 +1,123 @@
+		}
+	}
+	return diagnostics, true, nil
+}
+
+// attempts to find which context in the config might be a cluster-admin for the server in the current context.
+func (o DiagnosticsOptions) findClusterClients(rawConfig *clientcmdapi.Config) (*client.Client, *kclient.Client, bool, error) {
+	if o.ClientClusterContext != "" { // user has specified cluster context to use
+		if context, exists := rawConfig.Contexts[o.ClientClusterContext]; exists {
+			configErr := fmt.Errorf("Specified '%s' as cluster-admin context, but it was not found in your client configuration.", o.ClientClusterContext)
+			o.Logger.Error("discClustCtx", configErr.Error())
+			return nil, nil, false, configErr
+		} else if os, kube, found, err := o.makeClusterClients(rawConfig, o.ClientClusterContext, context); found {
+			return os, kube, true, err
+		} else {
+			return nil, nil, false, err
+		}
+	}
+	currentContext, exists := rawConfig.Contexts[rawConfig.CurrentContext]
+	if !exists { // config specified cluster admin context that doesn't exist; complain and quit
+		configErr := fmt.Errorf("Current context '%s' not found in client configuration; will not attempt cluster diagnostics.", rawConfig.CurrentContext)
+		o.Logger.Errorf("discClustCtx", configErr.Error())
+		return nil, nil, false, configErr
+	}
+	// check if current context is already cluster admin
+	if os, kube, found, err := o.makeClusterClients(rawConfig, rawConfig.CurrentContext, currentContext); found {
+		return os, kube, true, err
+	}
+	// otherwise, for convenience, search for a context with the same server but with the system:admin user
+	for name, context := range rawConfig.Contexts {
+		if context.Cluster == currentContext.Cluster && name != rawConfig.CurrentContext && strings.HasPrefix(context.AuthInfo, "system:admin/") {
+			if os, kube, found, err := o.makeClusterClients(rawConfig, name, context); found {
+				return os, kube, true, err
+			} else {
+				return nil, nil, false, err // don't try more than one such context, they'll probably fail the same
+			}
+		}
+	}
+	return nil, nil, false, nil
+}
+
+// makes the client from the specified context and determines whether it is a cluster-admin.
+func (o DiagnosticsOptions) makeClusterClients(rawConfig *clientcmdapi.Config, contextName string, context *clientcmdapi.Context) (*client.Client, *kclient.Client, bool, error) {
+	overrides := &clientcmd.ConfigOverrides{Context: *context}
+	clientConfig := clientcmd.NewDefaultClientConfig(*rawConfig, overrides)
+	factory := osclientcmd.NewFactory(clientConfig)
+	o.Logger.Debugf("discClustCtxStart", "Checking if context is cluster-admin: '%s'", contextName)
+	if osClient, kubeClient, err := factory.Clients(); err != nil {
+		o.Logger.Debugf("discClustCtx", "Error creating client for context '%s':\n%v", contextName, err)
+		return nil, nil, false, nil
+	} else {
+		subjectAccessReview := authorizationapi.SubjectAccessReview{
+			// we assume if you can list nodes, you're the cluster admin.
```

Comment:
not a good assumption, lots of things can list nodes... other nodes, for example. options:
1. check for something only a cluster admin should have access to, like `list oauthaccesstokens`
2. check for a bogus verb on a bogus noun, that will only match the `* on *` rule that cluster admins have, like `educate dolphins`